store/download_test.go

```diff
@@ -435,20 +441,136 @@ func (s *downloadSuite) TestUseDeltas(c *C) {
-		} else {
-			os.Setenv("PATH", altPath)
+			// just mock the xdelta3 command directly
+			hostXdelta3Cmd = testutil.MockCommand(c, "xdelta3", "")
 		}
 
-		c.Check(store.UseDeltas(), Equals, scenario.wantDelta, Commentf("%#v", scenario))
+		// if there is not meant to be xdelta3 on the host or in core, then set
+		// PATH to be empty such that we won't find xdelta3 from the host
+		// running these tests
+		if !scenario.exeInHost && !scenario.exeInCore {
+			os.Setenv("PATH", "")
+
+			// also reset PATH at the end, otherwise an empty PATH leads
+			// testutil.MockCommand fails in future iterations that mock a
+			// command
+			cleanups = append(cleanups, func() {
+				os.Setenv("PATH", origPath)
+			})
+		}
+
+		// run the check for delta usage, we call it twice
+		sto := &store.Store{}
+		c.Check(sto.UseDeltas(), Equals, scenario.wantDelta, comment)
+
+		// cleanup the files we may have created before calling the function
+		// again to ensure that the caching works as expected
+		if scenario.exeInCore {
+			err := os.Remove(interpInCorePath)
+			c.Assert(err, IsNil)
+		}
+
+		if scenario.exeInHost {
+			hostXdelta3Cmd.Restore()
+		}
+
+		// also now that we have deleted the mock interpreter and unset the
+		// search path, we should still get the same result as above when
+		// we call UseDeltas() since it was cached, if it wasn't cached then
+		// this would fail
+		c.Check(sto.UseDeltas(), Equals, scenario.wantDelta, comment)
+
+		if scenario.wantDelta {
+			// if we should have been able to use deltas, make sure we picked
+			// the expected one, - if both were true we should have picked the
+			// one from core instead of the one from the host first
+			if scenario.exeInCore {
+				// check that during trying to check whether to use deltas or
+				// not, we called the interpreter with the xdelta3 config
+				// command too
+				c.Check(coreInterpCmd.Calls(), DeepEquals, [][]string{
+					{"ld-linux-x86-64.so.2", "--library-path", "/some/dir/from/etc/ld.so", exeInCorePath, "config"},
+				}, comment)
+
+				// also check that now after caching the xdelta3 command, it
+				// returns the expected format
+				expArgs := []string{
+					interpInCorePath,
+					"--library-path",
+					"/some/dir/from/etc/ld.so",
+					exeInCorePath,
+					"foo",
+					"bar",
+				}
+				// check that the Xdelta3Cmd function we cached uses the
+				// interpreter that was returned from CommandFromSystemSnap
+				c.Check(sto.Xdelta3Cmd("foo", "bar").Args, DeepEquals, expArgs, comment)
+
+			} else if scenario.exeInHost {
+				// similar checks for the host case, except in the host case we
+				// just called xdelta3 directly
+				c.Check(hostXdelta3Cmd.Calls(), DeepEquals, [][]string{
+					{"xdelta3", "config"},
+				}, comment)
+
+				// and args are passed to the command cached too
+				expArgs := []string{hostXdelta3Cmd.Exe(), "foo", "bar"}
+				c.Check(sto.Xdelta3Cmd("foo", "bar").Args, DeepEquals, expArgs, comment)
+			}
+		}
```

Comment:
it feels like we should sanity check in the else here,  that both exe are false or env is false?

Reply:
added a check to this